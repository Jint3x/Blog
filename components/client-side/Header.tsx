import React, { useEffect, useState } from "react"


function Header() {
    let [isVisible, setIsVisible]: [any, any] = useState("inital");

    useEffect(() => {
        window.addEventListener("scroll", () => scrollFunction())

        // Used to detect when the page scroll is below 50px, if so, some new styles will be applied to the header.
        function scrollFunction() {
            if ((document.documentElement.scrollTop || document.body.scrollTop) > 75 && isVisible) {
                setIsVisible(false);
            } else if ((document.documentElement.scrollTop || document.body.scrollTop) < 75 && !isVisible) {
                setIsVisible(true)
            }
        }

        return (
            window.removeEventListener("scroll", scrollFunction)
        )
    }, [])

    return (
        <nav>
            <div className={`
            bg-black flex pt-1 pb-2 content-center flex-col fixed w-full transition-all transform duration-1000 z-50
            ${isVisible !== false ? "" : "-translate-y-10 sm:-translate-y-0"} sm:flex-row sm:pt-1 sm:h-10
            ${isVisible === "initial" ? "" : isVisible === true ? "sm:animate-headerResizeOpposite" : "sm:animate-headerResize"}
            `}>
                <LogoText />
                <Navigation />
            </div>
        </nav>
    )
}

function LogoText() {
    return (
        <div className={`w-full text-center sm:w-40 pl-3 pb-3`}>
                <p className={"text-2xl"}>
                    <span className={"text-main-blue relative bottom-2"}>Jint3x</span>
                    <span className={"text-main-green relative left-1"}>Blogs</span>
                </p>
        </div>
    )
}


function Navigation() {
    return (
        <div className={"w-full self-center sm:w-72 sm:pt-1 sm:ml-auto sm:mr-6"}>
                <ul className={"flex justify-around w-full"}>
                    <NavigationLink 
                     href={"/"}
                     text={"Home"}
                    />

                    <NavigationLink 
                     href={"/blogs"}
                     text={"Blogs"}
                    />

                    <NavigationLink 
                     href={"https://devjint3x.com"}
                     text={"About me"}
                    />
                </ul>
        </div>
    )
}

interface NavigationLink_I {
    text: string,
    href: string,
}

function NavigationLink(props: NavigationLink_I) {
    let [url, setURL] = useState("");

    useEffect(() => {
        setURL(window?.location?.pathname || "")
    })

    let normalNavigation = "text-main-blue"
    let selectedNavigation = "text-main-green bg-main-blue"

    return (
        <li className={`border-main-green border rounded w-20 text-center ${url === props.href ? selectedNavigation : normalNavigation} hover:bg-main-blue-lighter transition-all cursor-pointer`}>
            <a
             href={props.href}
             className={"w-20"}
            >
                {props.text}
            </a>
        </li>
    )
}



export default Header