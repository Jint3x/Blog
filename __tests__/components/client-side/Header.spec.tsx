/**
 * @jest-environment jsdom
 */
 import Header from "../../../components/client-side/Header";
 import ReactRenderer from "react-test-renderer"
 import { screen, render } from "@testing-library/react"


test("Renders the header and verifies its links", () => {
    render(<Header />)

    let blogs = screen.getAllByText("Blogs")[1] as HTMLAnchorElement;
    let home = screen.getByText("Home") as HTMLAnchorElement;
    let about = screen.getByText("About me") as HTMLAnchorElement;

    expect(blogs.nodeName).toBe("A");
    expect(blogs.href).toBe("http://localhost/blogs");

    expect(home.nodeName).toBe("A");
    expect(home.href).toBe("http://localhost/");

    expect(about.nodeName).toBe("A");
    expect(about.href).toBe("https://jint3x.dev/");
})


test("Matches the correct styles on the components", () => {
    render(<Header />)

    let selectedClasses = ["text-main-green", "bg-main-blue"];
    let normalClasses = ["text-main-blue"]; 

    let homepageClasses = Object.values(screen.getByText("Home")!.parentElement!.classList);
    let blogsClasses = Object.values(screen.getAllByText("Blogs")[1]!.parentElement!.classList);

    expect(window.location.pathname).toBe("/");
    expect(selectedClasses.every(className => homepageClasses.includes(className))).toBeTruthy()
    expect(normalClasses.every(className => blogsClasses.includes(className))).toBeTruthy()
})


test("Changes the header style based on the scroll position", () => {
    Object.defineProperty(window, "innerWidth", {
        writable: true,
        configurable: true,
        value: 400,
    });

    window.dispatchEvent(new Event("resize"));

    const { container } = render(<Header />);
    let mainDivBeforeScroll = (container.querySelector("nav")?.firstChild?.cloneNode() as HTMLDivElement)

    document.documentElement.scrollTop = 100;
    window.dispatchEvent(new Event("scroll"));

    let mainDivAfterScroll = container.querySelector("nav")?.firstChild as HTMLDivElement;

    expect(mainDivBeforeScroll.classList).not.toContain("-translate-y-10");
    expect(mainDivBeforeScroll.classList).not.toContain("sm:-translate-y-0");
    expect(mainDivAfterScroll.classList).toContain("-translate-y-10");
    expect(mainDivAfterScroll.classList).toContain("sm:-translate-y-0");
})
    


test("Takes a snapshot of the component", () => {
    let tree = ReactRenderer.create(
        <Header />
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
})
