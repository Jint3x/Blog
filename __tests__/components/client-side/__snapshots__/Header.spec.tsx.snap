// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Takes a snapshot of the component 1`] = `
<nav>
  <div
    className="
            bg-black flex pt-1 pb-2 content-center flex-col fixed w-full transition-all transform duration-1000 z-50
             sm:flex-row sm:pt-1 sm:h-10
            sm:animate-headerResize
            "
  >
    <div
      className="w-full text-center sm:w-40 pl-3 pb-3"
    >
      <p
        className="text-2xl"
      >
        <span
          className="text-main-blue relative bottom-2"
        >
          Jint3x
        </span>
        <span
          className="text-main-green relative left-1"
        >
          Blogs
        </span>
      </p>
    </div>
    <div
      className="w-full self-center sm:w-72 sm:pt-1 sm:ml-auto sm:mr-6"
    >
      <ul
        className="flex justify-around w-full"
      >
        <li
          className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
        >
          <a
            className="w-20"
            href="/"
          >
            Home
          </a>
        </li>
        <li
          className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
        >
          <a
            className="w-20"
            href="/blogs"
          >
            Blogs
          </a>
        </li>
        <li
          className="border-main-green border rounded w-20 text-center text-main-blue hover:bg-main-blue-lighter transition-all cursor-pointer"
        >
          <a
            className="w-20"
            href="https://devjint3x.com"
          >
            About me
          </a>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
